src/mds/MDCache.cc

```diff
@@ -9811,7 +9811,25 @@ void MDCache::request_forward(const MDRequestRef& mdr, mds_rank_t who, int port)
 void MDCache::dispatch_request(const MDRequestRef& mdr)
 {
   if (mdr->dead) {
-    dout(20) << __func__ << ": dead " << *mdr << dendl;
+    dout(10) << __func__ << ": dead " << *mdr << dendl;
+    //if the mdr is a "batch_op" and it has followers, pick a follower as
+    //the new "head of the batch ops" and go on processing the new one.
+    if (mdr->client_request && mdr->is_batch_head()) {
+      int mask = mdr->client_request->head.args.getattr.mask;
+      auto it = mdr->batch_op_map->find(mask);
+      auto new_batch_head = it->second->find_new_head();
+      if (!new_batch_head) {
+        mdr->batch_op_map->erase(it);
+        dout(10) << __func__ << ": mask '" << mask
+                 << "' batch head is dead and there is no follower" << dendl;
+        return;
+      }
+      dout(10) << __func__ << ": mask '" << mask
+               << "' batch head is dead and queue a new one "
+               << *new_batch_head << dendl;
+      mds->finisher->queue(new C_MDS_RetryRequest(this, new_batch_head));
+      return;
+    }
     return;
```

Comment:
Slightly off topic but related question - If the mds is dead and is not a batch op, it's just returned here. So what's the lifecycle for this op here ?

Reply:
If it's not a batch head req, so it shouldn't be here. And the lifecycle will be end just when the corresponding session was killed and the session's requests were killed and cleaned as below:

```
2024-04-16T10:39:58.181327563Z debug 2024-04-16T10:39:58.180+0000 7f5b98c56640 10 mds.0.server kill_session 0x562fa774aa00
2024-04-16T10:39:58.181346482Z debug 2024-04-16T10:39:58.180+0000 7f5b98c56640 10 mds.0.server journal_close_session : client.188978 10.128.2.216:0/1255716376 pending_prealloc_inos [] free_prealloc_inos [] delegated_inos []
2024-04-16T10:39:58.181358961Z debug 2024-04-16T10:39:58.180+0000 7f5b98c56640 20 mds.0.sessionmap mark_projected s=0x562fa774aa00 name=client.188978 pv=737759 -> 737760
2024-04-16T10:39:58.181374045Z debug 2024-04-16T10:39:58.180+0000 7f5b98c56640 10 mds.0.cache request_kill request(client.188978:3 nref=4 cr=0x562fa71e3500) 
2024-04-16T10:39:58.181389776Z debug 2024-04-16T10:39:58.180+0000 7f5b98c56640 15 mds.0.cache request_cleanup request(client.188978:3 nref=4 cr=0x562fa71e3500) 
```

But for the batch head request, whose `mdr->nref` will be increased when making it as a batch head, since the `mdr->nref` is not `1` so the lifecycle won't end by marking as `mdr->dead=true`. 

So when retrying the batch head req it will just return without releasing the `mdr->nref` from the batch head struct. This is why just after the client was evicted the slow reqeust number decreased to `1`, but later it increased to 3 or more again, and the new ones were all `lookup` requests and will be treated as `followers` and added to the stale batch head.

Reply:
Thanks for the explaination